scheduler/src/cook/check_image.clj

```diff
@@ -0,0 +1,147 @@
+      ; TODO: What to return on other outputs? XXXX thinks we should default fail.
+      :else {:status :ok
+             :cache-expire-at (t/plus now odd-result-cache-timeout) })))
+
+(defn process-response-for-image-deployment-cache
+  "Process a response containing a json body with a map with two keys:
+    'built' and 'deployed' and true/valse as values."
+  [docker-image {:keys [docker-image status body]}]
+  (let [now (t/now)
+        built? ("built" body)
+        deployed? ("deployed" body)
+        ready? (and built? deployed?)]
+    (cond
+      (and (= 200 status) ready?) {:status :ok
+                                   :cache-expire-at (t/plus now good-cache-timeout) }
+      (and (= 200 status) (not ready?)) {:status :later
+                                         :cache-expire-at (t/plus now unknown-cache-timeout) }
+      (= 404 status) {:status :error
+                      :message (str "Image " docker-image " not found")
+                      :cache-expire-at (t/plus now bad-cache-timeout)}
+      ; TODO: What to return on other outputs? XXXX thinks we should default fail.
+      :else {:status :ok
+             :cache-expire-at (t/plus now odd-result-cache-timeout) })))
+
+(defn generate-url-from-image
+  [docker-image]
+  "TODO")
+
+(defn image-deployment-miss
+  "Is the image deployed? Do the HTTP and return the response."
+  [docker-image]
+  (let [now (t/now)]
+    (try
+      (let [url (generate-url-from-image docker-image)
+            response (do-image-status-query-http url image-deployment-check-http-timeout-millis)]
+        (process-response-for-image-validity-cache docker-image response))
+      (catch Exception e
+        {:status :later
+         :cache-expire-at (t/plus now odd-result-cache-timeout)}))))
+
+(def ^LoadingCache image-deployment-cache (-> (CacheBuilder/newBuilder)
+                                              (.maximumSize 200000)
+                                              (.expireAfterAccess 72 TimeUnit/HOURS)
+                                              (.build (CacheLoader/from
+                                                        (reify Function
+                                                          (apply [_ docker-image]
+                                                            (image-deployment-miss docker-image)))))))
+
+(defn image-validity-miss [docker-image]
+  (try
+    (let [now (t/now)
+          url (generate-url-from-image docker-image)
+          response (do-image-status-query-http url image-validity-check-http-timeout-millis)
+          validity (process-response-for-image-validity-cache docker-image response)
+          {:keys [status] :as deployment} (process-response-for-image-deployment-cache docker-image response)]
+      ; Opportunistically put the status into the image-deployment-cache if its good.
+      (when (= :ok status)
+        (.put image-deployment-cache docker-image deployment))
```

Comment:
My initial reaction is that I don't like one miss function to be cascading writes to other caches. But I haven't thought through alternatives yet.